src/aiodynamo/client.py

```diff
@@ -417,67 +497,56 @@ def table(self, name: TableName) -> Table:
             select = Select.specific_attributes
         if select is Select.count:
             raise TypeError("Cannot use Select.count with query, use count instead")
 
-        expression_attribute_names = {}
-        expression_attribute_values = {}
+        params = Parameters()
 
-        projection_expression, ean = get_projection(projection)
-        expression_attribute_names.update(ean)
+        payload = {
+            "TableName": table,
+            "KeyConditionExpression": key_condition.encode(params),
+            "ScanForward": scan_forward,
+        }
 
-        builder = ConditionExpressionBuilder()
+        if projection:
+            payload["ProjectionExpression"] = projection.encode(params)
 
         if filter_expression:
-            filter_expression, ean, eav = builder.build_expression(filter_expression)
-            expression_attribute_names.update(ean)
-            expression_attribute_values.update(eav)
-
-        if key_condition:
-            key_condition_expression, ean, eav = builder.build_expression(
-                key_condition, True
-            )
-            expression_attribute_names.update(ean)
-            expression_attribute_values.update(eav)
-        else:
-            key_condition_expression = None
-
-        coro_func = partial(
-            self.core.query,
-            **clean(
-                TableName=table,
-                IndexName=index,
-                ScanIndexForward=scan_forward,
-                ProjectionExpression=projection_expression,
-                FilterExpression=filter_expression,
-                KeyConditionExpression=key_condition_expression,
-                ExpressionAttributeNames=expression_attribute_names,
-                ExpressionAttributeValues=py2dy(expression_attribute_values),
-                Select=select.value,
-            )
-        )
-        async for raw in unroll(
-            coro_func,
-            "ExclusiveStartKey",
-            "LastEvaluatedKey",
-            "Items",
-            py2dy(start_key) if start_key else None,
-            limit,
-            "Limit",
-        ):
-            yield dy2py(raw, self.numeric_type)
+            payload["FilterExpression"] = filter_expression.encode(params)
+
+        if start_key:
+            payload["ExclusiveStartKey"] = start_key
+        if index:
+            payload["IndexName"] = index
+        if limit:
+            payload["Limit"] = limit
+        if select:
+            payload["Select"] = select.value
+
+        payload["ExpressionAttributeNames"] = params.get_expression_names()
+        payload["ExpressionAttributeValues"] = params.get_expression_values()
+
+        got = 0
+
+        async for result in self._fast_depaginate("Query", payload):
+            for item in result["Items"]:
+                yield dy2py(item, self.numeric_type)
+                if limit:
+                    got += 1
+                    if got >= limit:
+                        return
```

Comment:
might as well preventively slice `result["Items"]` up to the limit.
somehow the code like `for ...: x += 1` just nags me.

Reply:
it's ugly, I agreee. but there's multiple `result['Items']`, so slicing wouldn't be that straight forward.

Ideally `_fast_depaginate` would deal with limits but then I'd have to have a separate one for `count` or the API of `_fast_depaginate` becomes very convoluted.